balancer/state.go

```diff
@@ -0,0 +1,165 @@
+package balancer
+
+import (
+	"context"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	radio "github.com/R-a-dio/valkyrie"
+	"github.com/R-a-dio/valkyrie/config"
+	"github.com/R-a-dio/valkyrie/errors"
+)
+
+// Balancer represents the state of the load balancer.
+type Balancer struct {
+	config.Config
+	manager radio.ManagerService
+	storage radio.StorageService
+	serv    *http.Server
+
+	// The current stream to re-direct clients to.
+	current atomic.Value
+
+	// The amount of listeners from every relay.
+	listeners int
+}
+
+func (br *Balancer) getCurrent() string {
+	return br.current.Load().(string)
+}
+
+// we know statically that current will only be of type string.
+func (br *Balancer) setCurrent(str string) {
+	br.current.Store(str)
+}
+
+// health checks the status of r using c, returning a copy of r.
+func health(ctx context.Context, c *http.Client, r radio.Relay) radio.Relay {
+	res := r
+	res.Online, res.Listeners = false, 0
+	req, err := http.NewRequestWithContext(ctx, "GET", r.Status, nil)
+	if err != nil {
+		return res
+	}
+	resp, err := c.Do(req)
+	if err != nil {
+		return res
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return res
+	}
+	resp.Body.Close()
+	l, err := parsexml(body)
+	if err != nil {
+		return res
+	}
+	res.Online, res.Listeners = true, l
+	return res
+}
+
+// checker receives relays on in and sends updated relays on out.
+// the execution of checker can be halted by cancelling ctx.
+func checker(ctx context.Context, in, out chan radio.Relay) {
+	c := &http.Client{
+		Timeout: 3 * time.Second,
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case relay, ok := <-in:
+			if ok {
+				out <- health(ctx, c, relay)
+			} else { // we've received every value and the channel is closed
+				close(out) // we're not sending anymore
+				return
+			}
+		}
+	}
+}
+
+// update checks all relays and sets the current relay to re-direct to.
+func (br *Balancer) update(ctx context.Context) {
+	relays, err := br.storage.Relay(ctx).All()
+	if err != nil {
+		if errors.Is(errors.NoRelays, err) {
```

Comment:
maybe log that no relays were found, might be easier to debug if it somehow ever happens